### ISSUE TITLE
%0A gets stripped from URL

### ISSUE THREAD

**Issue by adigitalife** [YOURLS/YOURLS#2056]
- YOURLS version:  Latest master

It seems that %0A (line feed) is always stripped from the URL when adding a new short link.  It doesn't matter if I disable encoding using the plugin in  #1651.  Is there any way to keep %0A (or any other control character) as is?

**Comment by LeoColomb** [comment]
X-Ref: #1303

**Comment by adigitalife** [comment]
In #1303, %0A would be converted to %250A, correct?  
For example: hello%0Aworld => hello%250Aworld

But what is happening is that %0A is being stripped out.
For example:  hello%0Aworld => helloworld

If %0A becomes %250A then I know how to fix it.  But if %0A disappears then what do I do?

**Comment by lordmilutin** [comment]
To "fix" (read avoid this)
I have found in 
`/includes/functions-formatting.php`
```
$strip = array( '%0d', '%0a', '%0D', '%0A' );
$url = yourls_deep_replace( $strip, $url );
```
Fix is easy, to comment that out, but since it's there on purpose, question is why is it there? Is there any security issue maybe or something similar?

**Comment by ozh** [comment]
Because it's the "proper way" of doing thing: encode non trivial characters into their encoded form. Eg linefeed become %0A, space becomes %20 and so on. See issue 1303 as linked above for more explanations.

**Comment by nebojsac** [comment]
@ozh The question was not about encoding - he's asking why the encoded characters were being stripped. They were encoded in the first place, so I assume they're safe after that.

**Comment by ozh** [comment]
The question is about encoding and stripping.

The root of the problem we still have to tackle is to be able to determine if a %0A (or similar) is intentional before encoding or a result of URL encoding. For instance, does the user wants to shorten `http://example.com/with space` (which will be rightfully converted as `with%20space`), or does he want to shorten `http://example.com/?ihavea%20inmystring` which will be converted as `%2520`.

**Comment by adigitalife** [comment]
@lordmilutin Good find!  Yes, this solves the problem.

@ozh I have the same question about the following code:
`$strip = array( '%0d', '%0a', '%0D', '%0A' );
$url = yourls_deep_replace( $strip, $url );`

This bit specifically addresses %0A and %0D (nothing to do with %20 or any other).  And specifically, it strips all the %0A and %0D from the long URLs.  So the question is, why remove these two in particular?

**Comment by ozh** [comment]
Hmmm. I'm sure I had a smart idea at that time :) Will think about this over the week-end. Maybe I overthought something.

**Comment by ozh** [comment]
Should be fixed by 6748ce8a947e38e5b4
